### DISCUSSION TITLE
Ionization

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#469] [General]
Let's start with the simplest case with the laser irradiation without any injected particles etc which could be of interest for everyone. Now when major bugs are found and hopefully fixed (and the last bug in any software will never be found no matter what the efforts  :)  )  how specifically will look the same demo input files like cone.deck from sample_decks in  epoch3d if we include collisional ionization ? Let's take as an example some middle z atom like nickel or whatever.  Case  can be simplified even more if we take only bottom of the cone which will model just the plane solid foil target.

On the output for ions each record for ions should have its ionization state Z shown. That is how some other PIC and hydro codes do, but i do not know how epoch is doing that. The electron output structure probably will be the same like in older versions.

**Comment by Status-Mirror** [comment]
Hey Dan,

Current documentation for switching on ionisation is present in a dedicated section on this page: 

```
https://epochpic.github.io/documentation/input_deck/input_deck_species.html
```

When we upgrade to the new version with the bug fix (which is currently under review by other developers), we will update the documentation on this page.

When you switch on ionisation, you must also add a key to species blocks to identify them as a species which can be ionised. When you specify its charge and atomic number, the ion is identified - electron binding energies are assigned from a look-up table, which sets the collisional ionisation cross sections. The code automatically generates different species for each ionisation state, so there's no ambiguity in which species is which.

*Currently a bug exists where if you manually specify the electron release species, the charges of the generated ion species are not set correctly*

Hope this helps,
Stuart

**Reply by DanRRRR** [reply]
Stuart,
I see this testing is not moving very fast. And I for example can not take part in this testing because not trying ionization even ones i really do not understand what is going on there,  how to specify ions correctly etc.. Other people i am sure also have no clue what is going on with testing without trying this before.  So if the other method, not the manual one, does not exhibit obvious bugs, i'd still encourage you to prepare such small test, if it would not take too much time. So we would continue testing the ionization in parallel, check the quality of the physical model, the run speed with different initial parameters etc. In our codes we have more than a half of a dozen of ionization models, its has its purpose and limits of validity. All periodically give us a headache, and one of them, the most advanced, even did not complete being exhausted with the issues :)

**Reply by Status-Mirror** [reply]
Hey Dan,

For EPOCH, all bug fixes must be reviewed by at least one other developer before going live, and the other developers are quite busy currently. There's not much I can do to speed the process along.

A description of our model is present in [this paper](https://aip.scitation.org/doi/pdf/10.1063/5.0126336). We have benchmarked the code by measuring the number of ions produced when mono-energetic electrons pass through a target, inferring a cross section, and demonstrating agreement with experimental collisional ionisation cross sections - see Figures 6 and 7.

If you're interested, I've provided one of the input decks used to create this benchmark (1D). For further ways to modify collisional ionisation parameters, see [here](https://epochpic.github.io/documentation/input_deck/input_deck_species.html).

```
begin:constant
  # User input
  electron_ke_ev = 480
  electron_ne = 2.5e24

  # Derived parameters
  electron_ke = electron_ke_ev * qe
  mc2 = me * c^2
  electron_momentum = sqrt((electron_ke + mc2)^2 - mc2^2) / c
  electron_speed = sqrt(1.0 - (electron_ke/mc2 + 1)^(-2)) * c
  
  # Physical parameters
  n_xi = 2.7e25
end:constant

begin:control
  nx = 6
  x_min = -2e-9
  x_max = 10e-9
  t_end = 1.5*(x_max - x_min) / electron_speed
  dt_multiplier = 0.8
  field_ionisation = F
  stdout_frequency = 50
end:control

begin:collisions
  use_collisions = T 
  collisional_ionisation = T
  collide = Electron Electron off 
end:collisions

begin:boundaries
  bc_x_min = simple_outflow
  bc_x_max = simple_outflow
end:boundaries

begin:species
  name = Xe
  charge = 0
  mass = 1836.2 * 131.0
  atomic_no = 54
  dump = T
  npart = 2000000
  rho = if (x gt 0, n_xi, 0)
  ionise = T
  ionise_limit = 1
  unique_electron_species = T
  immobile = T
end:species

begin:species
  name = Electron
  dump = T
  npart = 10000
  rho = if (x lt 0, electron_ne, 0)
  drift_x = electron_momentum
  identify:electron
end:species

begin:output
  dt_snapshot = t_end
  px = always
  py = always
  pz = always
  particle_weight = always
  number_density = always + species
end:output
```

**Comment by DanRRRR** [comment]
Thanks, Stuart. I have few questions about ionization after looking at this init.deck: 

If ionization takes care for number of generated electrons why in settings  npart = 10000 is used as well as  rho = if (x lt 0, electron_ne, 0) ?
If electrons are derivatives of density and ionization process of atoms, the  density of electrons should not be here, isn't it?

With "npart=10000" here is related question: it is set as a constant,but the ionization permanently creates electrons. Does this mean that the weight is permanently growing to compensate this growth while "npart" keeps to be a constant. Wouldn't it  be better always  use "particles_per_cell=some value" instead?

And why unique_electron_species are here? What is unique about them? Description is even more confusing. What is "ionisation_electron_species" ? Electrons do not ionize, ions do, but this keyword sounds exactly opposite.  The statement "Identify: electron" is also confusing, not clear what for it is used. I understand that it is very hard to invent short meaningful keywords though. 

I am having temporal difficulties to run this 1D epoch demo, currently i succeeded to see the data only with 3D epoch, or  i can also see 2D data but using Matlab (looks like Matlab m-files in SDF directory work only with 2D, the 3D SDF files just even do not load). Work is in progress, not all i currently can display from epoch, but with time all will work i hope.

**Reply by Status-Mirror** [reply]
I think there's some misunderstanding on what the collisional ionisation routine is doing. Maybe these points will help clear some things up:

- Parameters like `particles_per_cell`, `npart`, `frac` set the **initial** number of macro-particles to load into the simulation. The code does not force `npart` or `particles_per_cell` to remain constant over the simulation run-time.
- This simulation describes a short beam of electrons hitting a target and triggering collisional ionisation. The `Electron` species describes particles in the incident electron beam. When this collides with the target `Xe` species, it will trigger collisional ionisation.
- In a collisional ionisation event, there are 3 important particles: the target ion, the incident electron which triggers the ionisation, and the electron which is ejected from the ion after ionisation. 
- `identify:electron` is using the identify feature. EPOCH already knows the mass and charge of some particles, so using the identify means we don't need to specify these properties. It also flags the species as an electron for certain physics packages, including collisional ionisation. When we include this line, the code knows that particles in the Electron species have a mass $m_e$, charge $-e$, and that they can collide with atoms/ions to ionise them (provided these atom/ion species have the `ionise=T` flag).
  -  Incident electrons are in the Electron species
  - The `unique_electron_species` generates a second, separate species for electrons ejected from their ions after ionisation
  -  `ionisation_electron_species` can also be used to set a species to send ejected electrons to, but only for species already set in the input deck. If you wanted ejected eletrons to go to the `Electron` species instead of a new one, you would set `ionisation_electron_species = Electron`.
- In MATLAB, you need to make sure the .m files located in SDF/MATLAB are on your path. You can do this by rightclicking the folder in the MATLAB interface, and selecting "add to path". Alternatively, copy all these .m files to your working directory.

**Comment by DanRRRR** [comment]
To get some clues how ionization works here is the example i compiled from your demo above, Epoch Online Help about ionization, irradiation at the angle and 3D example cone.deck. Last one i used because it was really well done and worked from day one as opposed to many other decks for me ( i simplified it even more).  As you can see assembled new demo does not work complaining about something. Would appreciate if you look what is definitely wrong here. The only what needed from ionization in this example is charge number for each particle, just one more number like others weight, momentum etc.  

This is why epoch needs to collect considerably more demo examples to simplify its use. Setting variants with epoch is extremely hard work by some reason i still did not  figured out why. There are many examples on the net, but typically none of them work out of the box.
(Also how to post the code in this git? "Add code" does not work, using brackets <> or [code][/code], Crtl+E do not work too. Without that Comments if present go with gargantuan size )
```
begin:control
  nx = 250
  ny = 125
  nz = 125
  t_end = 500 * femto
  x_min = -15 * micron
  x_max = -x_min
  y_min = -25 * micron
  y_max = -y_min
  z_min =  y_min
  z_max =  y_max
  field_ionisation = F
end:control


begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = periodic
  bc_y_max = periodic
  bc_z_min = periodic
  bc_z_max = periodic
end:boundaries


begin:constant
  laser_wavelength = 8.e-7   
  omega = 2 * pi * c / laser_wavelength
  laser_k = 2 * pi / laser_wavelength
  den_solid= 12 * critical(omega)
  target_thickness = 1 * micron
  target_x_inner  = -target_thickness/2
  target_x_outer  =  target_thickness/2
  r = sqrt(y^2 + z^2)
  laser_angle = pi / 4
  laser_fwhm = 2.0e-6   

end:constant

begin:collisions
  use_collisions = T
  collisional_ionisation = T
  collide = electron electron off     #  fixed Electron Electron ---> electron electron.  Means elastic collisions off 
end:collisions

begin:species
  name = Silicon
  charge = 0.
  mass = 1836.2*28.09
  atomic_no = 14
  ionise = T
  npart_per_cell = 1
  unique_electron_species = T

   number_density = if(x lt target_x_inner, 0.0, den_solid)
   number_density = if(x gt target_x_outer, 0.0, den_solid)

end:species

begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  identify:electron
   npart_per_cell = 1
end:species


begin:laser
  boundary = x_min
  intensity_w_cm2 = 2.0e18
  profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))
  t_profile = 1
  phase = -y * laser_k * sin(laser_angle) # Vary wave-front position in space
  lambda = laser_wavelength * cos(laser_angle) # Space wave-fronts apart  # correctly along k

end:laser

begin:output_global
  force_final_to_be_restartable = T
end:output_global


begin:output
  name = normal

  
  dt_snapshot = 10 * femto

  grid = always
  ey = always
  average_particle_energy = always
  number_density = always + species
  temperature = always + species

 
  distribution_functions = always
end:output


begin:output
  name = large
  
  dt_snapshot = 20 * femto

   
  particles = always
  particle_weight = always
 
  charge_density = always
end:output


begin:dist_fn
  name = x_px
  ndims = 2

  direction1 = dir_x
  direction2 = dir_px
 
  range1 = (1, 1)
  range2 = (-3.0e-20, 3.0e-20)

   resolution1 = 1
  resolution2 = 100

  include_species:electron
  include_species:Silicon
end:dist_fn


begin:dist_fn
  name = x_px_py
  ndims = 3
  dumpmask = never

  direction1 = dir_x
  direction2 = dir_px
  direction3 = dir_py

   
  range1 = (1, 1)
  range2 = (-3.0e-20, 3.0e-20)
  range3 = (-3.0e-20, 3.0e-20)

   resolution1 = 1
  resolution2 = 100
  resolution3 = 100

  include_species:electron
  include_species:Silicon
end:dist_fn[/code]
```

**Reply by Status-Mirror** [reply]
This isn't working because species names are case sensitive. Your line: `collide = Electron Electron off #elastic collisions off ?` is switching off elastic Coulomb collisions between particles within the `Electron` species, when you actually define a species named `electron` instead (lowercase E). 

On your question, we did this as we didn't want Coulomb collisions to interfere with our measurement of the collisional ionisation cross section. This is a numerical quality benchmark, not a physical benchmark. We set the cross section in the code, and we want to see that the code produces the correct number of ions according to that cross section.

Finally, this GitHub comments section uses Markdown syntax. If you want to include a code, you have to write it like this:

\```
begin:something
  some_code = here
end:something
\```

to generate:

```
begin:something
  some_code = here
end:something
```

Cheers,
Stuart

**Reply by DanRRRR** [reply]
Thanks Stuart, after fixing capital letter the code started ok.
Wikipedia: Just one missing symbol in the software code caused 1962 Mariner1 rocket to crash, $0.7B in current money was lost 

Missing 1000 symbols happen to me everyday :)

**Reply by DanRRRR** [reply]
Still looks like there are some errors in this compiled from multiple sources deck. I am getting all energies = 0. Deck was compiled from 1D, 2D or 3D decks which are often not compatible, some were just incomplete small snippets of decks.

**Comment by DanRRRR** [comment]
Stuart & All others, Do anyone see other errors in this not working deck? It loads, produces something, but all energies are 0. Fields probably are ok (can not check that, my software does not like that particles are wrong, need to fix that. Also, I know how to plot this field using Matlab in 2D but will also appreciate a hint how to plot in Matlab the 3D field). Try it, i reduced arrays so it goes minutes on a PC

UPDATE. 
Found one minor not related to the subject error and fixed it in above deck. 
was : profile = gauss(y, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))
now : profile = gauss(r, 0, laser_fwhm / (2.0 * sqrt(loge(2.0))))

This is because i took this part from documentation but it probably was meant for 2D without mentioning that  - the 2D/3D settings mismatch will be EPOCH eternal problem. Some other PIC codes accept 2D and 3D settings without changes.  Documentation must stress 2D/3D differences in all its examples by warning users who will copy the demo snippets into their settings. Each such error will cause loss of days when you notice that because runs are slow.
This fix will not fix the zero energy of all particles.

**Comment by DanRRRR** [comment]
So far i have not succeeded to setup a single variant with any kind of ionization, even just the OFI which work. More examples are needed. May be special sections for that needed too.  Even demos not working. For example, does ionization.deck in epoch2d example_decks work? It stops immediately after step 1. No examples for 3D to find how ionization slows down the run. Ionization slows down the codes tremendously sometimes, so i am interested how epoch is doing this, a i saw it is very very fast without it. Can developers post the cone.deck demo3D with separately OFI and collisional ionization ionization for example starting with zero charge (neutral atoms)? Or how to setup initial ionization not with z=1,4 or 22 but with z=0.001 for example? And also show how it is possible to speed up the run doing different tweaks.  

/* the "cultural conflict" might be the reason for my fault. I am doing hydro and PIC simulations with ionization for decades (on different codes). But all of them  are setting ionization as well as many other things totally different way. This is why the documentation should illustrate each and every sentence containing the words "this can be done ..." because when i do what "can be done" my way -- this does not work.
